### PR TITLE
gh-112205: Support `@setter` annotation from AC

### PR DESCRIPTION
**Pull request by corona10**
* Issue: gh-112205

### CODE REVIEW

File: Tools/clinic/clinic.py [hunk 1]
```diff
@@ -1944,9 +1970,16 @@ def render_function(
...
-            template_dict['c_basename'] = f.c_basename + "_get"
+        if f.kind in (GETTER, SETTER):
+            template_dict['getsetter_name'] = f.c_basename.upper()
+            template_dict['getset_basename'] = f.c_basename
+            if f.kind is GETTER:
+                template_dict['c_basename'] = f.c_basename + "_get"
+            elif f.kind is SETTER:
+                template_dict['c_basename'] = f.c_basename + "_set"
+                # TODO: Improve handling pararms better way.
+                data.impl_parameters.append("PyObject *value")
```
Comment (corona10) [inline Tools/clinic/clinic.py:1982]: @erlend-aasland @AlexWaygood 
If you want to propose a better way for this, please let me know.

**Comment by corona10** [comment]
@colesbury @erlend-aasland @AlexWaygood 
The current implementation is not super pretty, so please leave feedback for this implementation.
I tried not to touch the overall architecture of AC as possible.

* Unify GET/SET method macro definition with macro under technique.
* Update AC to support `@setter` annotation.

File: Tools/clinic/clinic.py [hunk 2]
```diff
@@ -1959,7 +1993,10 @@ def render_function(
             converter.set_template_dict(template_dict)
 
         f.return_converter.render(f, data)
-        template_dict['impl_return_type'] = f.return_converter.type
+        if f.kind is SETTER:
+            template_dict['impl_return_type'] = 'int'
```
Comment (corona10) [inline Tools/clinic/clinic.py:1998]: https://github.com/python/cpython/pull/112922/files#r1421696103
ditto

**Comment by corona10** [comment]
For the dev guide, I will send a PR once the PR is landed.

**Comment by corona10** [comment]
And also, we need to consider supporting PyDoc for getter/setter with the separated PR.
https://github.com/python/cpython/blob/384d6c1fe7e24d1c88f85c3f443e5420e54ca935/Modules/_io/textio.c#L181-L186

File: Tools/clinic/clinic.py [hunk 3]
```diff
@@ -870,8 +874,20 @@ class CLanguage(Language):
             {{"{name}", {methoddef_cast}{c_basename}{methoddef_cast_end}, {methoddef_flags}, {c_basename}__doc__}},
     """)
     GETTERDEF_PROTOTYPE_DEFINE: Final[str] = normalize_snippet(r"""
-        #define {getter_name}    \
-            {{"{name}", (getter){c_basename}, NULL, NULL}},
+        #if defined({getsetter_name}_GETSETTERDEF)
+        #   undef {getsetter_name}_GETSETTERDEF
+        #   define {getsetter_name}_GETSETTERDEF {{"{name}", (getter){getset_basename}_get, (setter){getset_basename}_set, NULL}},
+        #else
+        #   define {getsetter_name}_GETSETTERDEF {{"{name}", (getter){getset_basename}_get, NULL, NULL}},
```
Comment (erlend-aasland) [inline Tools/clinic/clinic.py]:
```suggestion
        #if defined({getset_name}_GETSETDEF)
        #   undef {getset_name}_GETSETDEF
        #   define {getset_name}_GETSETDEF {{"{name}", (getter){getset_basename}_get, (setter){getset_basename}_set, NULL}},
        #else
        #   define {getset_name}_GETSETDEF {{"{name}", (getter){getset_basename}_get, NULL, NULL}},
```

File: Tools/clinic/clinic.py [hunk 4]
```diff
@@ -870,8 +874,20 @@ class CLanguage(Language):
...
+        #endif
+    """)
+    SETTERDEF_PROTOTYPE_DEFINE: Final[str] = normalize_snippet(r"""
+        #if defined({getsetter_name}_GETSETTERDEF)
+        #   undef {getsetter_name}_GETSETTERDEF
+        #   define {getsetter_name}_GETSETTERDEF {{"{name}", (getter){getset_basename}_get, (setter){getset_basename}_set, NULL}},
+        #else
+        #   define {getsetter_name}_GETSETTERDEF {{"{name}", NULL, (setter){getset_basename}_set, NULL}},
```
Comment (erlend-aasland) [inline Tools/clinic/clinic.py]:
```suggestion
        #if defined({getset_name}_GETSETDEF)
        #   undef {getset_name}_GETSETDEF
        #   define {getset_name}_GETSETDEF {{"{name}", (getter){getset_basename}_get, (setter){getset_basename}_set, NULL}},
        #else
        #   define {getset_name}_GETSETDEF {{"{name}", NULL, (setter){getset_basename}_set, NULL}},
```

File: Tools/clinic/clinic.py [hunk 5]
```diff
@@ -1226,12 +1246,19 @@ def parser_body(
             limited_capi = False
 
         parsearg: str | None
+        if f.kind in {GETTER, SETTER} and len(parameters) > 0:
+            raise RuntimeError("@getter, @setter should not define paramters")
```
Comment (erlend-aasland) [inline Tools/clinic/clinic.py]:
```suggestion
            fail("neither @getter nor @setter can define parameters")
```

File: Tools/clinic/clinic.py [hunk 6]
```diff
@@ -1944,9 +1971,16 @@ def render_function(
         full_name = f.full_name
         template_dict = {'full_name': full_name}
         template_dict['name'] = f.displayname
-        if f.kind is GETTER:
-            template_dict['getter_name'] = f.c_basename.upper() + "_GETTERDEF"
-            template_dict['c_basename'] = f.c_basename + "_get"
+        if f.kind in (GETTER, SETTER):
+            template_dict['getsetter_name'] = f.c_basename.upper()
```
Comment (erlend-aasland) [inline Tools/clinic/clinic.py]:
```suggestion
            template_dict['getset_name'] = f.c_basename.upper()
```

File: Tools/clinic/clinic.py [hunk 7]
```diff
@@ -1172,6 +1188,10 @@ def output_templates(
         elif f.kind is GETTER:
             methoddef_define = self.GETTERDEF_PROTOTYPE_DEFINE
             docstring_prototype = docstring_definition = ''
+        elif f.kind is SETTER:
+            return_value_declaration = "int return_value;"
```
Comment (erlend-aasland) [inline Tools/clinic/clinic.py]: The return value variable name has a template (`{return_value}`). It _should_ be possible to use it most places, but unfortunately, it is still hard-coded most places 😢 It should map correctly here, though:

```suggestion
            return_value_declaration = "int {return_value};"
```

File: Tools/clinic/clinic.py (hunk 1 above)
Comment (erlend-aasland) [reply Tools/clinic/clinic.py:1982]: I'm not sure it is a good idea to add the `value` param implicitly. OTOH, it does reduce clinic input boilerplate code; all setters have the same argument spec, so why not declare it implicitly?

Implicitly adding the param prevents being able to add a param docstring, or a custom param name. That _could_ be solved by introspection, but I really do not think the added complexity would be worth it.

Alternatively, you could add a check that requires the clinic input to explicitly declare the `value` parameter.

I think implicitly adding it is fine.

File: Tools/clinic/clinic.py [hunk 8]
```diff
@@ -1944,9 +1971,16 @@ def render_function(
...
-            template_dict['getter_name'] = f.c_basename.upper() + "_GETTERDEF"
-            template_dict['c_basename'] = f.c_basename + "_get"
+        if f.kind in (GETTER, SETTER):
+            template_dict['getsetter_name'] = f.c_basename.upper()
+            template_dict['getset_basename'] = f.c_basename
+            if f.kind is GETTER:
+                template_dict['c_basename'] = f.c_basename + "_get"
+            elif f.kind is SETTER:
+                template_dict['c_basename'] = f.c_basename + "_set"
+                # TODO: Improve handling pararms better way.
```
Comment (erlend-aasland) [inline Tools/clinic/clinic.py]:
```suggestion
                # Implicitly add the setter value parameter.
```

File: Tools/clinic/clinic.py (hunk 2 above)
Comment (erlend-aasland) [reply Tools/clinic/clinic.py:1998]: I think this is fine; we do similar stuff for `__init__` functions. Perhaps a comment could be helpful, though.

```suggestion
            # All setters return an int.
            template_dict['impl_return_type'] = 'int'
```

File: Tools/clinic/clinic.py [hunk 9]
```diff
@@ -5335,8 +5374,15 @@ def at_critical_section(self, *args: str) -> None:
         self.critical_section = True
 
     def at_getter(self) -> None:
+        if self.kind in (GETTER, SETTER):
+            fail("@getter and @setter can not be declared at once")
```
Comment (erlend-aasland) [inline Tools/clinic/clinic.py:5378]: How about one of these? (Ditto for the `fail(...)` below.)

```suggestion
            fail("only one of @getter or @setter can be used")
```

(I'm not super enthusiastic about my suggestion; perhaps @AlexWaygood can cook up something nicer 😃)

Comment (erlend-aasland) [inline Tools/clinic/clinic.py]: How about this? (Ditto for the `fail(...)` below.)

```suggestion
            fail("only one of @getter or @setter can be used")
```

(I'm not super enthusiastic about my suggestion; perhaps @AlexWaygood can cook up something nicer 😃)

**Review by erlend-aasland — COMMENTED** [review]
Overall this looks good to me; thanks!

I would prefer to use a naming aligned with the names already used in the C API (`tp_getset`, `PyGetSetDef`); that is, consistently use `getset` and `GETSET` instead of `getsetter` and `GETSETTER`.

File: Tools/clinic/clinic.py (hunk 1 above)
Comment (corona10) [reply Tools/clinic/clinic.py:1982]: Yeah, if the current implicit declaration is acceptable, I am fine.
But we may consider refactoring them at some moment, even if we declare it implicitly.

File: Tools/clinic/clinic.py [hunk 10]
```diff
@@ -870,8 +874,20 @@ class CLanguage(Language):
...
+        #   define {getset_name}_GETSETDEF {{"{name}", (getter){getset_basename}_get, NULL, NULL}},
+        #endif
+    """)
+    SETTERDEF_PROTOTYPE_DEFINE: Final[str] = normalize_snippet(r"""
+        #if defined({getset_name}_GETSETDEF)
+        #   undef {getset_name}_GETSETDEF
+        #   define {getset_name}_GETSETDEF {{"{name}", (getter){getset_basename}_get, (setter){getset_basename}_set, NULL}},
+        #else
+        #   define {getset_name}_GETSETTERDEF {{"{name}", NULL, (setter){getset_basename}_set, NULL}},
+        #endif
```
Comment (erlend-aasland) [inline Tools/clinic/clinic.py:890]: Nitpick: I see the other preprocessor templates are using either 2 or 4 space indents. Let's align with the existing style before merging.

```suggestion
        #  undef {getset_name}_GETSETTDEF
        #  define {getset_name}_GETSETDEF {{"{name}", (getter){getset_basename}_get, (setter){getset_basename}_set, NULL}},
        #else
        #  define {getset_name}_GETSETDEF {{"{name}", (getter){getset_basename}_get, NULL, NULL}},
        #endif
    """)
    SETTERDEF_PROTOTYPE_DEFINE: Final[str] = normalize_snippet("""
        #if defined({getset_name}_GETSETDEF)
        #  undef {getset_name}_GETSETDEF
        #  define {getset_name}_GETSETDEF {{"{name}", (getter){getset_basename}_get, (setter){getset_basename}_set, NULL}},
        #else
        #  define {getset_name}_GETSETTERDEF {{"{name}", NULL, (setter){getset_basename}_set, NULL}},
```

File: Tools/clinic/clinic.py [hunk 11]
```diff
@@ -870,8 +874,20 @@ class CLanguage(Language):
...
+        #else
+        #   define {getset_name}_GETSETDEF {{"{name}", (getter){getset_basename}_get, NULL, NULL}},
+        #endif
+    """)
+    SETTERDEF_PROTOTYPE_DEFINE: Final[str] = normalize_snippet(r"""
+        #if defined({getset_name}_GETSETDEF)
+        #   undef {getset_name}_GETSETDEF
+        #   define {getset_name}_GETSETDEF {{"{name}", (getter){getset_basename}_get, (setter){getset_basename}_set, NULL}},
+        #else
+        #   define {getset_name}_GETSETTERDEF {{"{name}", NULL, (setter){getset_basename}_set, NULL}},
```
Comment (erlend-aasland) [inline Tools/clinic/clinic.py]: Nitpick: I see the other preprocessor templates are using either 2 or 4 space indents. Let's align with the existing style before merging.

```suggestion
        #  undef {getset_name}_GETSETTDEF
        #  define {getset_name}_GETSETDEF {{"{name}", (getter){getset_basename}_get, (setter){getset_basename}_set, NULL}},
        #else
        #  define {getset_name}_GETSETDEF {{"{name}", (getter){getset_basename}_get, NULL, NULL}},
        #endif
    """)
    SETTERDEF_PROTOTYPE_DEFINE: Final[str] = normalize_snippet("""
        #if defined({getset_name}_GETSETDEF)
        #  undef {getset_name}_GETSETDEF
        #  define {getset_name}_GETSETDEF {{"{name}", (getter){getset_basename}_get, (setter){getset_basename}_set, NULL}},
        #else
        #  define {getset_name}_GETSETTERDEF {{"{name}", NULL, (setter){getset_basename}_set, NULL}},
```

File: Tools/clinic/clinic.py [hunk 12]
```diff
@@ -1226,12 +1246,19 @@ def parser_body(
             limited_capi = False
 
         parsearg: str | None
+        if f.kind in {GETTER, SETTER} and len(parameters) > 0:
```
Comment (erlend-aasland) [inline Tools/clinic/clinic.py]:
```suggestion
        if f.kind in {GETTER, SETTER} and parameters:
```

**Comment by corona10** [comment]
Thank you @erlend-aasland, I 've updated PR PTAL

**Comment by erlend-aasland** [comment]
Thanks. I'm ok with this PR, but let's wait for @AlexWaygood's review before we land it.

File: Tools/clinic/clinic.py [hunk 13]
```diff
@@ -870,8 +874,20 @@ class CLanguage(Language):
             {{"{name}", {methoddef_cast}{c_basename}{methoddef_cast_end}, {methoddef_flags}, {c_basename}__doc__}},
     """)
     GETTERDEF_PROTOTYPE_DEFINE: Final[str] = normalize_snippet(r"""
-        #define {getter_name}    \
-            {{"{name}", (getter){c_basename}, NULL, NULL}},
+        #if defined({getset_name}_GETSETDEF)
+        #  undef {getset_name}_GETSETTDEF
```
Comment (erlend-aasland) [inline Tools/clinic/clinic.py]:
```suggestion
        #  undef {getset_name}_GETSETDEF
```

File: Tools/clinic/clinic.py [hunk 14]
```diff
@@ -870,8 +874,20 @@ class CLanguage(Language):
...
+        #else
+        #  define {getset_name}_GETSETDEF {{"{name}", (getter){getset_basename}_get, NULL, NULL}},
+        #endif
+    """)
+    SETTERDEF_PROTOTYPE_DEFINE: Final[str] = normalize_snippet(r"""
+        #if defined({getset_name}_GETSETDEF)
+        #  undef {getset_name}_GETSETDEF
+        #  define {getset_name}_GETSETDEF {{"{name}", (getter){getset_basename}_get, (setter){getset_basename}_set, NULL}},
+        #else
+        #  define {getset_name}_GETSETTERDEF {{"{name}", NULL, (setter){getset_basename}_set, NULL}},
```
Comment (erlend-aasland) [inline Tools/clinic/clinic.py]:
```suggestion
        #  define {getset_name}_GETSETDEF {{"{name}", NULL, (setter){getset_basename}_set, NULL}},
```

**Review by erlend-aasland — COMMENTED** [review]
Oops, I probably introduced a typo with my recent suggestions; sorry 'bout that!

**Review by erlend-aasland — APPROVED** [review]
LGTM. Let's wait for Alex to chime in.

(cc. @colesbury, if you'd like to take a look.)

File: Tools/clinic/clinic.py [hunk 15]
```diff
@@ -1226,12 +1246,19 @@ def parser_body(
             limited_capi = False
 
         parsearg: str | None
+        if f.kind in {GETTER, SETTER} and parameters:
+            fail("neither @getter nor @setter can define parameters")
```
Comment (AlexWaygood) [inline Tools/clinic/clinic.py]: I think we could give a slightly nicer error message here:

```suggestion
            fail(f"@{f.kind.name.lower()} method cannot define parameters")
```

File: Tools/clinic/clinic.py [hunk 16]
```diff
@@ -1944,9 +1971,16 @@ def render_function(
         full_name = f.full_name
         template_dict = {'full_name': full_name}
         template_dict['name'] = f.displayname
-        if f.kind is GETTER:
-            template_dict['getter_name'] = f.c_basename.upper() + "_GETTERDEF"
-            template_dict['c_basename'] = f.c_basename + "_get"
+        if f.kind in (GETTER, SETTER):
```
Comment (AlexWaygood) [inline Tools/clinic/clinic.py]: (ignorable nit)

```suggestion
        if f.kind in {GETTER, SETTER}:
```

File: Tools/clinic/clinic.py [hunk 17]
```diff
@@ -5335,8 +5375,15 @@ def at_critical_section(self, *args: str) -> None:
         self.critical_section = True
 
     def at_getter(self) -> None:
+        if self.kind in (GETTER, SETTER):
+            fail("only one of @getter or @setter can be used")
```
Comment (AlexWaygood) [inline Tools/clinic/clinic.py]: From the user's perspective, I think these are actually two distinct errors, so it makes sense to give two different error messages:

```suggestion
        if self.kind is GETTER:
            fail("Cannot apply @getter twice to the same function!")
        if self.kind is SETTER:
            fail("Cannot apply both @getter and @setter to the same function!")
```

File: Tools/clinic/clinic.py [hunk 18]
```diff
@@ -5335,8 +5375,15 @@ def at_critical_section(self, *args: str) -> None:
...
         self.kind = GETTER
 
+    def at_setter(self) -> None:
+        if self.kind in (GETTER, SETTER):
+            fail("only one of @getter or @setter can be used")
```
Comment (AlexWaygood) [inline Tools/clinic/clinic.py]:
```suggestion
        if self.kind is SETTER:
            fail("Cannot apply @setter twice to the same function!")
        if self.kind is GETTER:
            fail("Cannot apply both @getter and @setter to the same function!")
```

File: Tools/clinic/clinic.py [hunk 19]
```diff
@@ -5536,6 +5583,8 @@ def state_modulename_name(self, line: str) -> None:
 
         return_converter = None
         if returns:
+            if self.kind in {GETTER, SETTER}:
+                fail("neither @getter nor @setter can define return type")
```
Comment (AlexWaygood) [inline Tools/clinic/clinic.py]:
```suggestion
                fail(f"@{self.kind.name.lower()} methods cannot define a return type")
```

**Review by AlexWaygood — COMMENTED** [review]
Thanks! A few small suggestions, mostly to do with error messages:

File: Tools/clinic/clinic.py (hunk 17 above)
Comment (AlexWaygood) [reply Tools/clinic/clinic.py]: It also looks like this failure path might not have any test coverage; could you add a test for it? :)

File: Tools/clinic/clinic.py (hunk 18 above)
Comment (AlexWaygood) [reply Tools/clinic/clinic.py]: It also looks like this failure path might not have any test coverage; could you add a test for it? :)

File: Tools/clinic/clinic.py [hunk 20]
```diff
@@ -5335,7 +5375,22 @@ def at_critical_section(self, *args: str) -> None:
...
+            case FunctionKind.GETTER:
+                fail("Cannot apply @getter twice to the same function!")
+            case FunctionKind.SETTER:
+                fail("Cannot apply both @getter and @setter to the same function!")
+            case _:
+                self.kind = FunctionKind.GETTER
+
+    def at_setter(self) -> None:
+        match self.kind:
+            case FunctionKind.SETTER:
```
Comment (corona10) [inline Tools/clinic/clinic.py:5388]: @AlexWaygood How about this?

Comment (AlexWaygood) [reply Tools/clinic/clinic.py:5388]: Very nice!

File: Lib/test/test_clinic.py [hunk 21]
```diff
@@ -2197,6 +2197,56 @@ class Foo "" ""
...
+                    module foo
+                    class Foo "" ""
+                    {annotation}
+                    {annotation}
+                    Foo.property -> int
+                """
+                expected_error = f"Cannot apply {annotation} twice to the same function!"
+                self.expect_failure(block, expected_error, lineno=3)
+
+        dup_annotations = [("@getter", "@setter"), ("@setter", "@getter")]
```
Comment (AlexWaygood) [inline Lib/test/test_clinic.py:2239]: I think it makes sense to have this as a separate test method:

```suggestion
    def test_getter_and_setter_disallowed_on_same_function(self):
        dup_annotations = [("@getter", "@setter"), ("@setter", "@getter")]
```

File: Lib/test/test_clinic.py [hunk 22]
```diff
@@ -2197,6 +2197,56 @@ class Foo "" ""
...
+        for dup in dup_annotations:
+            with self.subTest(dup=dup):
+                block = f"""
+                    module foo
+                    class Foo "" ""
+                    {dup[0]}
+                    {dup[1]}
+                    Foo.property -> int
+                """
+                expected_error = f"Cannot apply both @getter and @setter to the same function!"
```
Comment (AlexWaygood) [inline Lib/test/test_clinic.py]: To fix the lint failure:

```suggestion
                expected_error = "Cannot apply both @getter and @setter to the same function!"
```

File: Lib/test/test_clinic.py [hunk 23]
```diff
@@ -2197,6 +2197,57 @@ class Foo "" ""
...
+
+                block = f"""
+                   module foo
+                   class Foo "" ""
+                   {annotation}
+                   Foo.property
+                       obj: int
+                       /
+                """
+                expected_error = f"{annotation} method cannot define parameters"
```
Comment (AlexWaygood) [inline Lib/test/test_clinic.py:2221]: It doesn't look like the `expected_error` variable here is actually used; did you mean to do this? :)

```suggestion
                expected_error = f"{annotation} method cannot define parameters"
                self.expect_failure(block, expected_error)
```

Comment (corona10) [reply Lib/test/test_clinic.py:2221]: Thank you for catching it!

**Review by AlexWaygood — APPROVED** [review]
LGTM, thanks!

**Pull request merged**